docs(mesh): add ECS installation (#3678)
Kong/docs.konghq.com: app/mesh/1.6.x/installation/ecs.md

---
title: Kong Mesh on Amazon ECS
---

This page describes running Kong Mesh on ECS and offers guidelines
for integrating Kong Mesh into your deployment process.

For inspiration, we have an [example repository of Cloudformation](https://github.com/Kong/kong-mesh-ecs)
that covers bootstrapping an ECS cluster from scratch, deploying Kong Mesh
and deploying some services into the mesh.

## Overview

On ECS, Kong Mesh runs in Universal mode. Every ECS task runs with an Envoy sidecar.
Kong Mesh supports tasks on the following launch types:

- Fargate
- EC2

The control plane itself also runs as an ECS service in the cluster.

### Dataplane tokens

As part of joining and synchronizing with the mesh, every sidecar needs to authenticate with
the control plane. On Kong Mesh, this is accomplished by using a dataplane token.

> Typically on Universal mode, creating and managing Dataplane tokens is a manual step for the mesh operator.

Author: changed this to be "data plane token" according to a discussion our team had around how to write these things consistently.

tasks to communicate with Kong Mesh the control plane.

The example Cloudformation includes two Cloudformation stacks for [creating a
cluster](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/vpc.yaml) and
[deploying Kong Mesh](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/controlplane.yaml)

### Kong Mesh ECS controller

The controller is published as a docker image
`docker.io/kong/kong-mesh-ecs-controller:0.1.0`.

#### API permissions

In order to generate dataplane tokens, the controller
needs to authenticate with the Kong Mesh API and be authorized to create
new dataplane tokens.

Our example repository [launches the control plane with two additional containers](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/controlplane.yaml#L358-L387)
that handle fetching this global secret and
[covers bootstrapping this controller and running it as an ECS task](https://github.com/Kong/kong-mesh-ecs/blob/main/README.md#ecs-controller).

Any option that enables operators to query the API from `localhost` (e.g. an SSH
container in the task) will allow extracting the admin token.

Once we have `kumactl` configured with this token, we can generate a new user

> token for the controller with a command like:

Author: The idea isn't to have users go back and forth or follow along at all. This section is essentially the instructions on how to use the ECS controller, which is in a private repo.
Any references to the example repo are nothing more than examples of how to do the things on this page, nothing to do with following along.

If anything, I can probably reduce the number of references to the example repo.

```
kumactl generate user-token --name ecs-controller --group mesh-system:admin --valid-for 8766h
```

We can configure our controller with this token using:

- the `KUMA_API_TOKEN` environment variable

We recommend communicating with the Kong Mesh API over TLS (served on port `5682` by default) and the controller supports:

- the `KUMA_API_CA_BYTES` environment variable for passing the CA used to verify the TLS certificates presented by the API

#### IAM permissions

The controller uses the AWS API and its ECS task role must be authorized to perform:

- `ecs:ListTasks` and `ecs:DescribeTasks`
- `secretsmanager:GetSecretValue` and `secretsmanager:PutSecretValue`

These permissions can of course be further restricted by including a `Resource` or `Condition` in

> the IAM policy statements. To make this easier, the controller supports:
> 
> - a `--secret-name-prefix` command line switch to prefix the names of the
>   AWS secrets under which it saves tokens

Reviewer: ```suggestion
the IAM policy statements. To make this easier, the controller supports a `--secret-name-prefix` command line switch to prefix the names of the AWS secrets under which it saves tokens.
```

> See [the example controller Cloudformation](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/controller.yaml) for how all of this ties together.

Reviewer: ```suggestion
To see how all of this ties together, refer back to the [example controller Cloudformation](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/controller.yaml).
```

## Services

When deploying an ECS task to be included in the mesh, the following must be
considered.

### Outbounds

> Like on Universal, services are bootstrapped with a `Dataplane` specification.

Reviewer: ```suggestion
Services are bootstrapped with a `Dataplane` specification.
```

> Transparent proxy is not supported on ECS and so the `Dataplane` resources for a

Reviewer: ```suggestion
Transparent proxy is not supported on ECS, so the `Dataplane` resource for a
```

> service must enumerate all other mesh services that service contacts and include them
> [in `Dataplane` specification as `outbounds`](https://kuma.io/docs/1.4.x/documentation/dps-and-data-model/#dataplane-specification).

Reviewer: ```suggestion
service must enumerate all other mesh services this service contacts and include them
[in the `Dataplane` specification as `outbounds`](https://kuma.io/docs/latest/documentation/dps-and-data-model/#dataplane-specification).
```
"That" causes ambiguous meaning.

> See the example repository for [how to handle the Dataplane template
> with Cloudformation](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/counter-demo/demo-app.yaml#L30-L46).

Reviewer: ```suggestion
See the example repository to learn
[how to handle the Dataplane template with Cloudformation](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/counter-demo/demo-app.yaml#L30-L46).
```

### `kuma.io/service` tag

> Every ECS task must be tagged with the `kuma.io/service` tag in
> order for the controller to include the task in the mesh. The ECS task

Reviewer: ```suggestion
Every ECS task must be tagged with the `kuma.io/service` tag so that the controller
includes the task in the mesh. The ECS task
```

> authenticates to the mesh as this service and it should match the
> `kuma.io/service` value in the `Dataplane` resource.

Reviewer: ```suggestion
authenticates to the mesh as this service. The task name should match the
`kuma.io/service` value in the `Dataplane` resource.
```
What exactly must match? Is it a name value? Is it something else?

### Sidecar

> The sidecar must run as a container in the ECS task and be passed the AWS secret
> that holds the dataplane token created by the ECS controller.

Reviewer: ```suggestion
The sidecar must run as a container in the ECS task, and it must receive the AWS secret
that holds the dataplane token created by the ECS controller.
```
This reads better, but I can understand if the term has to be "passed".

Author: tweaked it a bit differently into two sentences and "it must run with the AWS secret"

> The controller _does not create_ the secret, it only puts and gets it. The

Reviewer: ```suggestion
The controller _does not create_ the secret, it only PUTs and GETs it. The
```
If this is referring to requests, then capitalize the terms.

Author: hmm it's referring to AWS API actions

Reviewer: So it's a GET or a PUT request from the AWS API? Or is it something else?

Author: It's the PutSecretValue and GetSecretValue actions which don't map to HTTP methods

Reviewer: Got it. Feel free to ignore, then :)

AWS secret should be created and destroyed by the same mechanism that creates the

> ECS service (e.g. a Cloudformation stack).

Reviewer: ```suggestion
ECS service (for example, a Cloudformation stack).
```

See the example repository for [an example container
definition](https://github.com/Kong/kong-mesh-ecs/blob/main/deploy/counter-demo/demo-app.yaml#L205-L243).

#### Initialization

When a task starts, the following happens:

1. The task requests the `token` JSON key from an existing AWS secret.

> 1. Initially, the secret does not contain this key and ECS will continue

Reviewer: ```suggestion
1. Initially, the secret does not contain this key and ECS continues
```
